### PR TITLE
Improvements to expr sorting, various changes from norm_hack.

### PR DESCRIPTION
**Pull request by csarofeen**
In some more advanced uses during normalization expr sorting had issues resolving. Primary change there is to recompute the producer and compute at domains from scratch when merging, instead of trying to incrementally resolving them. It's not a big change, but fixes later use cases when we pull scheduling in.

Also changes to disable unrolling by env variable for RNG based tests.
Minor fix for trivial reduction.
Minor fix on symbolic size replacement.
Minor cleanup.

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/lower_loops.cpp [hunk 1]
```diff
@@ -108,12 +108,11 @@ void LoopNestGenerator::handle(const Expr* expr) {
...
-    // If out_id is derived from trivial reductions and its root axes
-    // are also all the case, it's safe to skip this axis.
-    if (gpu_lower->trivialReductionInfo().isDerivedFromRoot(out_id)) {
-      continue;
-    }
+    // Note: It is not safe to skip trivial reduction axes as they could be
+    // inlined with other tensor views. This happens in
+    // NVFuserTest.FusionBNRepro_CUDA as of this commit on norm_hack_2_rebased
+    // branch
+
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_loops.cpp:115]: This is fine and should be done so if anything breaks. However, if it is actually safe to skip, this change would mean we would generate unnecessary loops.

For example, here are excerpts from the `FusionTrivialReduction1` test with and without this change.

Current:
```
__global__ void kernel1(Tensor<float, 1> T0, Tensor<float, 1> T2, Tensor<float, 1> T8, Tensor<float, 1> T11, Tensor<float, 3> T3) {
  #pragma unroll 1
  for(size_t ki136 = 0; ki136 < T0.size[0]; ++ki136) {
    float T9[1];
    T9[0]
       = T0[(ki136 * T0.stride[0])];
    float T10[1];
    T10[0]
       = T9[0];
    T11[(ki136 * 1)]
       = T10[0];
  }
``` 

With this change:
```
__global__ void kernel1(Tensor<float, 1> T0, Tensor<float, 1> T2, Tensor<float, 1> T8, Tensor<float, 1> T11, Tensor<float, 3> T3) {
  #pragma unroll 1
  for(size_t ki142 = 0; ki142 < T0.size[0]; ++ki142) {
    float T9[1];
    T9[0]
       = T0[(ki142 * T0.stride[0])];
    float T10[1];
    #pragma unroll 1
    for(size_t ki150 = 0; ki150 < (ceilDiv(1, 4)); ++ki150) {
      #pragma unroll 1
      for(size_t ki152 = 0; ki152 < 4; ++ki152) {
        T10[0]
           = T9[0];
      }
    }
    #pragma unroll 1
    for(size_t ki144 = 0; ki144 < (ceilDiv(1, 5)); ++ki144) {
      #pragma unroll 1
      for(size_t ki146 = 0; ki146 < 5; ++ki146) {
        T11[(ki142 * 1)]
           = T10[0];
      }
    }
  }
```

The compiler should be able to optimize away the trivial loops in this case, but that might depend on other factors.

It should be also possible to skip loop generation only when no other tensor is computed-at. I'll create an issue to revisit this if desired for performance reasons.

File: torch/csrc/jit/codegen/cuda/kernel_cache.cpp [hunk 2]
```diff
@@ -95,6 +95,7 @@ at::DimVector graphReductionAxes(
 
   at::DimVector reduction_axes;
   // TODO: let check that we have only single reduction node in the graph.
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/kernel_cache.cpp:97]: nitpick: this comment above is not relevant any more, let's remove it.

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 3]
```diff
@@ -111,19 +111,9 @@ class IrParser {
...
-    // with eager mode
-    bool disable_unroll = false;
-    bool has_reduction = false;
     // compose nodes in topo order;
     for (const JitOp* node : block->nodes()) {
       processJitNode(node);
-      if (node->kind() == aten::rand_like) {
-        disable_unroll = true;
-      }
-      if (node->kind() == aten::sum) {
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/parser.cpp:124]: Nice catch!

**Review by jjsjann123 — APPROVED** [review]
LGTM

File: torch/csrc/jit/codegen/cuda/kernel_cache.cpp (hunk 2 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/kernel_cache.cpp:97]: Will do tomorrow in a follow up, merging :-P

**Pull request merged**